mut/index/SnootyManifest.py

```diff
@@ -0,0 +1,215 @@
+from unittest.result import STDERR_LINE
+from bson import decode_all
+from jsonpath_ng.ext import parse
+from os import walk
+from os.path import join
+from json import dumps
+from datetime import datetime
+from concurrent.futures import ProcessPoolExecutor
+
+from typing import Any, Dict, Optional, List
+
+class Document:
+    '''Return indexing data from a page's AST for search purposes.'''
+    def __init__(self, data) -> None:
+
+        self.tree = data[0]
+
+        self.robots, self.keywords, self.description = self.findMetadata()
+
+        self.paragraphs = self.findParagraphs()
+        self.code = self.findCode()
+        self.title, self.headings = self.findHeadings()
+        self.slug = self.deriveSlug()
+        self.preview = self.derivePreview()
+
+        self.noindex, self.reasons = self.get_noindex()
+
+    def findParagraphs(self) -> str:
+        logger("\tFinding paragraphs")
+        # NB: paragraphs include "paragraph" nodes within tables
+        jsonpath_expr = parse('$..children[?(@.type==\'paragraph\')]..value')
+        results = jsonpath_expr.find(self.tree)
+        
+        # Appending to then joining an array is faster than repeatedly concatenating strings
+        str_list = []
+        for r in results:
+            str_list.append(r.value)
+            
+        return ' '.join(str_list)
+
+    def findCode(self):
+        logger("\tFinding code")
+        jsonpath_expr = parse('$..children[?(@.type==\'code\')]')
+        results = jsonpath_expr.find(self.tree)
+        code_contents = []
+        for r in results:
+            lang = r.value.get("lang", None)
+            code_contents.append({"lang": lang, "value": r.value["value"]})
+        
+        return code_contents
+
+    def findHeadings(self):
```

Comment:
the return type of this appears to be `Tuple[Optional[str], Optional[List[str]]`.